### PR TITLE
Issue #7 | welcoming prompt for empty list - rb & ad

### LINKED ISSUE #7: 7. As a user, I want to see a welcoming prompt to add my first item if my list is empty to help me get oriented to how the app works.
## Summary

When a user starts a new list, they need a little guidance on how to get started. Think through what you'd want to see as a user and make some intentional user experience decisions that will guide our users in confidently adding their very first item. 

## Acceptance criteria
- [x] The list view, when there are no items to display, should show a prompt (e.g., a button) for the user to add their first item

## Notes
- [The wireframe](https://github.com/the-collab-lab/smart-shopping-list/blob/main/_resources/wireframes/list-empty.png) is a little short on the "welcoming" part of the story, but otherwise is functionally complete for the purposes of the AC.

### PR DESCRIPTION
**Pull request by Bustamor**
## Description

**Code we changed:**

In `App.jsx` we added a loading state that is set to `false` when the data has been pulled from Firestore. This `loading` value can then be passed as a prop into other components, as we did in `List.jsx`. In the return of `List.jsx` there is a ternary operator checking if the `loading` state is true or not, and another encasing ternary operator testing if the shopping list is empty or filled.

In `List.jsx`, a ternary operator checks for a boolean value from `checkForEmptyList()`, the user will then see one of two things:  
- If no items are in list, the user will see a prompt to add their first item with the use of a button below.  This will take them to the `Add Item` page.  
- Alternatively, If items are already in the list, the filter input and list of the existing items will render below.

The `checkForEmptyList()` function is how the shopping list data is filtered and checked for length. The `filter()` method is used to create a new array filled with elements that have a name property. We have to filter the items because new collections created in Firestore automatically have one unidentified empty id. The new array with a correct amount of shopping items is checked for a length of 0 and returns true if the list is empty, and false otherwise.

**Rationale**:
These view and functionality options were based on whether the user was joining a pre-existing list or beginning a new list.  

**What we learned:**

Several things came to light as we worked on this feature.  One adaptation we had to make was to base the ternary operator on what firebase gives us.  We had to account for `data.length` being 1 as the empty list still contains an id that we have to account for.  
Further, we had to think about what message the user would recieve that corresponds the the absence or presence of a list.  Some minimal styling of an icon button helped us understand how to attach images in the public folder and place them in the `List.jsx` view. 
## Related Issue

Closes #7 

## Acceptance Criteria

- [x] The list view, when there are no items to display, should show a prompt (e.g., a button) for the user to add their first item
## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓   | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

### Before
Before #7

### After
After #7

## Testing Steps / QA Criteria

- From your terminal, first run `git pull` and then `git checkout rb-ad-welcoming-prompt-for-empty-list`
- Next, run `npm start` to open http://localhost:3000/
- You can either access an empty shopping list by using the respective share token  or click on the button _Create a new list_ to view the new feature
- From there you will be redirected to the new list view that has a welcoming add item prompt for an empty list

### CODE REVIEW

File: src/views/List.jsx [hunk 1]
```diff
@@ -22,7 +26,22 @@ export function List({ data }) {
 		setFilterInput('');
 	};
 
-	return (
+	const handleAddItem = () => {
+		/* navigate to add item view */
+		navigate('/add-item');
+	};
+
+	return data.length == 1 ? (
```
Comment (sarahmjenkins) [inline src/views/List.jsx]: Would this work for `my test list` where there isn't an extra empty document in the collection? That may be a unique enough scenario that it may not matter here but just curious if there would be cases where there is one item that isn't the empty one.

**Review by sarahmjenkins — APPROVED** [review]
Good job, @Bustamor and @alex-andria !! I like the way you decided to do this one. I also like the cute little shopping cart icon you added and hope you have plans to use that somewhere :)

File: src/views/List.jsx (hunk 1 above)
Comment (arcan9) [inline src/views/List.jsx]: At first I was wondering why an empty list is equivalent to `data.length === 1` here. But I realized it was because every time a list is created an empty document is automatically generated. Like Sarah, I wonder how to make it so that no empty document is generated upon creation of the list.

**Review by arcan9 — APPROVED** [review]

File: src/views/List.jsx (hunk 1 above)
Comment (mtfoley) [reply src/views/List.jsx]: I think you should consider "sanitizing" the data state value upstream (i.e. filtering data array to be only non-empty items). This way you could use the data.length === 0 case.

**Review by mtfoley — APPROVED** [review]
Great job @Bustamor and @alex-andria! I chimed in on the topic of the the `data.length === 1` case in the comments.

**Review by djtaylor8 — APPROVED** [review]
Great work @alex-andria and @Bustamor! Just have a couple questions...

**Comment by djtaylor8** [comment]
@Bustamor/ @alex-andria  are you planning to use the icon you imported as part of this PR?

**Comment by djtaylor8** [comment]
Also, this might be a trivial thing, but I noticed that when I click on create new list, it looks like the list view flashes briefly then shows the welcome message. I wonder if there is a way to address that?

https://user-images.githubusercontent.com/78327387/216682357-6bcb90af-4b04-45d1-981a-9d1cdfa6bb0a.mov

**Review by daredtech — APPROVED** [review]

File: src/views/List.jsx (hunk 1 above)
Comment (Bustamor) [reply src/views/List.jsx]: @mtfoley Hey Matt. Thanks for the suggestions.  Could you clarify  a bit more what you mean by sanitizing upstream?  We've been experimenting, but no dice yet.  Thanks!

Comment (alex-andria) [reply src/views/List.jsx]: Randall and I worked on it for a few hours earlier, but I'm now using map and includes and I'm getting somewhere!

Comment (alex-andria) [reply src/views/List.jsx]: Good point! I'm not sure if this is something to be updated in how we make collections in `firebase.js`, but we fixed it in `List.jsx` so that whether or not an extra empty document is created in a new list collection, our logic should work. 

We updated the code so that it filters the values of a collection shopping list and then checks if it contains items that has the name property.

Comment (alex-andria) [reply src/views/List.jsx]: This is the route we ended up going with, feel free to let us know if there's any changes we should make!

`	/* This function will return a boolean value of whether the list
	is empty or not */
	const checkForEmptyList = () => {
		//return true if list is empty and false if list contains items
		const shoppingListArr = data.map((item) => item.name);
		return shoppingListArr.includes(undefined) ? true : false;
	};`

**Comment by alex-andria** [comment]
> Also, this might be a trivial thing, but I noticed that when I click on create new list, it looks like the list view flashes briefly then shows the welcome message. I wonder if there is a way to address that?
> 
>  Screen.Recording.2023-02-03.at.10.39.11.AM.mov

Thank you for your comments @djtaylor8! To address your comments:

- The shopping cart icon was initially used, but we decided not to include it in our code because we wanted our group to have a clean slate when deciding styling later
- We addressed the list view flashing with a loading view that waits half a second before loading the list view

 `	useEffect(() => {
		setTimeout(() => {
			setLoading(false);
		}, 500);
	}, []);`

**Comment by djtaylor8** [comment]
@alex-andria I think having a loading state is definitely on the right track! However, since we're dealing with a network request here (getting a list/items from Firestore), hardcoding a value in for a time could potentially lead to some weirdness. I don't think this is a blocking issue right now, but I wonder if the loading state could be added somewhere else. I suggest checking out where the data is retrieved (`App.jsx`) and maybe creating a loading state there that can be passed down as a prop.

File: src/App.jsx [hunk 2]
```diff
@@ -40,6 +42,9 @@ export function App() {
 
 			/** Finally, we update our React state. */
 			setData(nextData);
+			/* Once data is pulled, the loading state will be set to
```
Comment (djtaylor8) [inline src/App.jsx:45]: 🤌🤌🤌🤌🤌🤌🤌

File: src/views/List.jsx [hunk 3]
```diff
@@ -1,12 +1,25 @@
...
-	/* Use handler to change the state of filterInput  and convert all items to lowercase to facilitate a more thorough search */
+	/* Declare navigate for view redirection */
+	const navigate = useNavigate();
+
+	/* This function will return a boolean value of whether the list
+	is empty or not */
+	const checkForEmptyList = () => {
+		/* return true if list is empty and false if list contains items */
+		const shoppingListArr = data.map((item) => item.name);
+		return shoppingListArr.includes(undefined) ? true : false;
```
Comment (djtaylor8) [inline src/views/List.jsx]: @alex-andria / @Bustamor this will evaluate to `true` since the blank item will return undefined here

Comment (alex-andria) [reply src/views/List.jsx]: Good catch! Resolved with `filter()` and then a length check later instead of includes.

File: src/views/List.jsx [hunk 4]
```diff
@@ -22,36 +36,57 @@ export function List({ data }) {
...
-					  })}
-			</ul>
-		</>
-	);
+	/*Handles navigation to Add Item view */
+	const handleAddItem = () => {
+		navigate('/add-item');
+	};
+
+	if (loading) {
```
Comment (djtaylor8) [inline src/views/List.jsx:45]: We can create a new issue here to look into adding loading spinner or something similar. Not needed right now.

**Review by alex-andria — APPROVED** [review]
